### PR TITLE
Google Appengine Datastore Backend

### PR DESCRIPTION
**Pull request by panamafrancis**
Opening a pull request for issues: 
https://github.com/google/cayley/issues/178
&& partially
https://github.com/google/cayley/issues/144

I've finished the functionality for the appengine datastore backend and got it running again.

At the moment the go appengine SDK is still using go 1.2 and although i've been informed that change is coming, i've yet to hear anything concrete.  This means that i'm not using sync pools yet. 

I haven't had the chance to do any load testing yet such as the '30kmoviedata', this is due to the instability of goapp test, it falls over when you try and add more than 100 entities at a time (socket errors), i did the relevant optimizations to my machine but it's still not enough, I believe my unit tests cover enough for now.

So I'm in need of a code review.

If you want to test it:
`cd graph\gaedatastore`
`goapp test -logtostderr`

if you want to test the HTTP server with gaedatastore backend, then
`cd appengine\`
`goapp serve`

### CODE REVIEW

**Comment by bashtian** [comment]
I added memcache support and fixed some bugs, perfomance is not bad. I did not test the 30kmoviedata yet, but works for me on the server with a small dataset. 
https://github.com/bashtian/cayley/tree/memcache

**Comment by panamafrancis** [comment]
Hey the whole 'Goon' integration is a great idea, I wouldn't have thought of that. I need to rebase the branch with the master and also change the way primary keys are implemented so that we can use UUID's.

I've got two tiny issues, first there is a lot of commented out code and debugging helpers if you could clean that up that would be great. Second the way you're working on the repository is kinda wrong, so if you take a look at your commit "Changed import path" [1], so instead of cloning your fork of the cayley repo it's best to clone the 'google/cayley' repo and then add your 'bastian/cayley' fork as a new remote, this way you don't have to change any import paths. 

[1] https://github.com/bashtian/cayley/commit/58278051297d01414281a1bd0639e722e03f1efa

Currently Go 1.4 is still beta for app engine, this is kinda holding everything back.

**Comment by bashtian** [comment]
I changed the import path to have a separate version in my gopath, I will revert it when I make the pull request. But I'm not sure how the workflow is since I rebased all of your commits to the latest master (at this time). I don't want to open a new pull request for that, but if that is the best option I can do it.

**Comment by panamafrancis** [comment]
So @bashtian  i would recommend you fork my branch and make a pull request on that.  then if all is well i'll merge your PR into this PR and so on. 

so your remotes would look like so:
origin (google/cayley)
panamafrancisfork
bashtianfork

**Comment by SeanDolphin** [comment]
appengine is now at 1.9.18 running go 1.4.1 so sync pools should be available.

**Comment by panamafrancis** [comment]
Right you are! Cheers.

**Comment by barakmich** [comment]
After taking the proper time to review (and thank you for writing really good tests) I'm merging this (by hand, thank you Github).

I made the few nit-worthy changes myself as part of the non-automatic merge, mostly in PrimaryKey JSON marshalling taking way fewer bytes for the common case of integers.

At first I questioned the GetHandleForRequest context wrapper, but in honesty, it's kind of the hack that was required. That one's on me for an HTTP router refactor someday, so I approve it given the fact that it does what it can and is minimally invasive.

Sure there may be some rollback questions, but then I realized that they all happened in the same HTTP request context. So that makes the story even stronger. 

In short, it's pretty damn good. Feel free to send a followup PR adding yourself to A+C. You deserve recognition!

LGTM

**Comment by panamafrancis** [comment]
Ahh I thought I'd made a total faux pas! :D

I've got a couple of extra changes to push first if you could wait just two
more days.

But thanks for the kind words.

Tschüss
Stefan

**Comment by barakmich** [comment]
Rebase on master (cause I just merged it all) and send them in a follow-up, along with A+C. 

Nah, the delay was all on me. No faux pas here :)
